### ISSUE TITLE
CSS modules duplicated when used with `next/dynamic`

### ISSUE THREAD

**Issue by mrmckeb** [vercel/next.js#25456]
### What version of Next.js are you using?

10.2.2

### What version of Node.js are you using?

14.16.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

When using `next/dynamic` with CSS modules, we've found that CSS can be re-inserted (duplicated)

In our scenario, we have a simple `<Text />` component that is already in use throughout the application. When we used `<Text />` in a new component that was dynamically imported, styles for `<Text />` were re-inserted.

These styles were inserted below all other styles, breaking some styles that rely on specificity - which is how we found this bug.

As a temporary workaround, we've removed that component from the dynamically loaded component.

### Expected Behavior

Styles should not be re-inserted/duplicated.

### To Reproduce

We haven't been able to reliably reproduce this in a standalone repo.

We've looked and found some potentially similar issues, and this _might_ be related: https://github.com/vercel/next.js/issues/20203.

**Comment by sgehrman** [comment]
I'm getting duplicate css without using dynamic.

**Comment by timneutkens** [comment]
@sgehrman please provide a reproduction

**Comment by bscaspar** [comment]
I am encountering a similar issue - CSS modules are getting duplicated and overwritten for similar components on a page - e.g. implementations of a custom `<Button />` component - after one of these components is dynamically loaded. No issue when the conditional dynamic component is not loaded. There's a clear flash of changing styles once the dynamic component is loaded.

**Comment by mrmckeb** [comment]
This has been resolved by #33032.

PR #33047 is also related, but has an additional ask.

**Comment by vborodulin** [comment]
any updates on this issue?

**Comment by vborodulin** [comment]
still have issue

**Comment by rikani** [comment]
Next.js `v12.1.0` still has problem.

**Comment by harisraharjo** [comment]
Not sure if I have the exact same issue but I'm also experiencing duplicated css (without dynamic import). 
In my case I'm using sprinkles from vanilla-extract to create my own tailwind (utility classes). By default webpack will create a standalone css file for the utility classes, and the other files that extend from it will also have their own file, which is nice. But I notice that if the size of the utility classes is less than 14.4 KB, webpack will copy **everything** from inside the utility classes and for **every** file that extends the utility classes will be injected with it. 
Is it possible to create seperate chunks without size limit? @timneutkens

**Comment by Puvvl** [comment]
Next.js `12.1.4` still has problem

**Comment by ycfreeman** [comment]
12.1.6 still has problem

**Comment by PhilipeGatis** [comment]
12.1.5 still has problem

**Comment by NikooDev** [comment]
12.1.6 still has problem : 
<img width="316" alt="Capture d’écran 2022-06-09 à 00 04 51" src="https://user-images.githubusercontent.com/43614012/172725481-3b9642f8-6df1-435b-9158-f5f19278fa1b.png">

**Comment by PepeBotella25** [comment]
My guess is that loader's outcome is being cached by webpack then when following chunks get compiled cached source is used causing the class names duplication. 
Dunno if it's possible to tell webpack to do not cache certain files

**Comment by PepeBotella25** [comment]
Repo to reproduce this:
https://github.com/PepeBotella25/css_module_dynamic_issue

Initially
![image](https://user-images.githubusercontent.com/46162931/177164863-0331b682-998f-4075-90c6-facc62dbac68.png)

After adding dynamic component using the same "shared" component
![image](https://user-images.githubusercontent.com/46162931/177164912-c78b8c61-027a-470f-8b4c-d998d21d613d.png)


Since you need css chunks to reproduce it you have to do build+start
```
npm run build
npm run start
```

**Comment by X7Becka** [comment]
Yeah, 12.2.3 still no workaround

**Comment by yakovlev-alexey** [comment]
Experienced this issue in 12.3.1 and found a workaround.

**TL;DR:** add a hidden (`display: none` should work) div in App/Page. In this div render all the components that have their styles duplicated.

---

Explanation: webpack (or whatever) seems to check whether components used in a `dynamic` import are already present on the page. In my case everything was fine when a loadable component was imported on a specific page. Later I moved the component to `_app.tsx` and styles started to duplicate in the CSS file emitted for the loadable component.

Say you have a few components:
- App (`pages/_app.tsx`)
- Page (`pages/index.tsx`)
- Button (`components/button.tsx`)
- LoadableModal (`components/loadable-modal.tsx`)
- LoadableModal uses Button

\* `uses` means imports either directly or indirectly but without dynamic imports.

The issue may be reproduced in a few ways:
1) LoadableModal is used in App, Button is used in Page **(sure, tested)**
2) LoadableModal is used in Page, Button is used in App **(not sure, need to test)**
3) Button is used neither in Page nor in App. There is an additional LoadableForm component that uses Button used on the same page. **(not sure, need to test)**

When building Next.js checks imports to see whether components used in `dynamic` are already present on the page. Since the relation between Page and App is not direct there is no way for it to detect that the Button is already used.

**My workaround was the following:** add a hidden (`display: none` should work) div in App/Page. In this div render all the components that have their styles duplicated. With this fix in place Next.js should properly deduplicate styles.

**Comment by aloeugene** [comment]
We're also experiencing this with Next@13.1.5.

<img width="460" alt="image" src="https://user-images.githubusercontent.com/21076597/217815902-60432686-502e-469e-b051-b20ec725f2e8.png">

And it's not reproducing running locally with `next dev`, which makes it even worse given the fact that we need to double-check such issues by building "prod" version locally OR just observing it when it gets to QA env (which uses "prod" version).

**Comment by Webbrother** [comment]
The same issue without using `dynamic`.
next@13.2.4

I'm wondering why this problem persists for a few years and a lot of tickets was closed without any solution.

- https://github.com/vercel/next.js/issues/11448
- https://github.com/vercel/next.js/issues/12343
- https://github.com/vercel/next.js/issues/11946
- https://github.com/vercel/next.js/issues/12099
- https://github.com/vercel/next.js/issues/20203

**Comment by potapovnikita** [comment]
any solutions for this issue?

**Comment by ananevandrei11** [comment]
Still not fixed `13.1.6`

**Comment by vasylkorchynskyi** [comment]
Any updates here?  Have the same issue with next@13.4.7


<img width="439" alt="image" src="https://github.com/vercel/next.js/assets/9628243/7479fce5-6602-4d6a-b68c-705757537e6e">

**Comment by AustinCase** [comment]
I am also having this issue with `next@13.4.5` -- I am actively trying to work on reproduction, perhaps will lend itself to the issue.

My instinct is that it has to do with SSR and the bundling magic between layouts/pages...

**Comment by roman-veryovkin** [comment]
I have the same issue with next@13.4.17. Any workarounds so far?

**Comment by AustinCase** [comment]
I can actively reproduce it without dynamic.
- Create a shared component outside of app directory.
- Create two or more pages that utilize the shared component.
- Notice that the shared component has duplicate styles.

Even if you have not visited both pages, if you have a Link component ( or something else that preloads the other page ). The server will stream down two css files, both with the EXCACT same selectors.

The browser for good reason can't dedup css selectors, instead it essentially does a merge. Ultimately I am just worried about the performance hit here, since it is doing the dedup on basically every shared component. Its also messy to read and a bad dev experience.

Hope this helps :)

**Comment by Adrastus** [comment]
We started a new project where we use CSS Modules. And the same story, styles are duplicated. By the way, large projects like @ https://github.com/payloadcms/payload have the same problem.

**Comment by ForsyteRun** [comment]
Any solution?

**Comment by monolithed** [comment]
This bug is [present](https://github.com/vercel/next.js/issues/51030) even without using next/dynamic

**Comment by maiconsanson** [comment]
Two years later, the issue persists... WTF!
Using `14.0.1`

<img width="733" alt="Screenshot 2023-11-02 at 11 07 16" src="https://github.com/vercel/next.js/assets/1965455/131ac4b7-b8b3-4eba-9a42-1c8e96a907c3">

**Comment by Egor-Dubovik** [comment]
I have the same issue in next@13.5.6 even without dynamic imports. Seriously this problem has been going on for so long and it still hasn't been fixed?

**Comment by zachsitka** [comment]
This is still an issue for me using next@14.1.3

**Comment by maiconsanson** [comment]
It seems to be solved in `14.1.3` without dynamic imports (production only, dev still duplicating).

https://github.com/vercel/next.js/pull/61198

**Comment by ivangorbunoff** [comment]
is there any solution? In large projects where components are often reused, duplicating styles greatly increases the final css file size

**Comment by Cirelion** [comment]
It's been over 3 years, how is there not a single update on this issue?

**Comment by lalexgap** [comment]
We still see duplicate CSS for dynamic imports on 14.2.3 in our production builds. It would be great to have a fix for this!

**Comment by muhammadzarif903** [comment]
To prevent CSS duplication when using next/dynamic, here's what you can do:

1. Import the CSS module locally in the component instead of globally. This way, the CSS will only apply to that specific component.

2. If needed, you can disable server-side rendering (SSR) for the dynamically loaded component by setting { ssr: false }.

3. Make sure the CSS modules are scoped to each component, so they don’t interfere with others or get duplicated.

**Comment by Juke-99** [comment]
I tested this issue with Next.js `16.3.3` (latest) and `16.4.0-canary.9` using the Pages Router and CSS Modules.

A shared component was rendered normally and was also used inside a component loaded with `next/dynamic`.

I tested the following configurations:

* Turbopack in development and production
* Webpack in development and production
* Next.js `16.3.3`
* Next.js `16.4.0-canary.9`

I could not reproduce the issue in any of these cases. Loading the dynamic component did not change the styles of the already rendered component, and I also could not see the shared CSS rule being inserted twice in DevTools.

There have also been significant changes to CSS chunking and ordering since this issue was reported, such as #63157 and #75049, which may be related to why this no longer reproduces.

Given that this no longer seems to be reproducible on current versions, I think it may be reasonable to close this issue.
